pkg/services/contexthandler/authproxy/authproxy.go

```diff
@@ -278,9 +278,27 @@ func (auth *AuthProxy) LoginViaHeader() (int64, error) {
 	}
 
 	auth.headersIterator(func(field string, header string) {
-		if field == "Groups" {
+		switch field {
+		case "Groups":
 			extUser.Groups = util.SplitString(header)
-		} else {
+		case "Role":
+			// If Role header is specified, we update the role of the specified org, or the default org
+			if header != "" {
+				rt := models.RoleType(header)
+				if rt.IsValid() {
+					extUser.OrgRoles = map[int64]models.RoleType{}
+					orgID := auth.orgID
+					if orgID == 0 {
+						if setting.AutoAssignOrg && setting.AutoAssignOrgId > 0 {
+							orgID = int64(setting.AutoAssignOrgId)
+						} else {
+							orgID = int64(1)
+						}
+					}
+					extUser.OrgRoles[orgID] = rt
```

Comment:
I think using `X-Grafana-Org-Id` here would allow a user to escalate their privilege in another org (say they're admin in org A and viewer in org B, they'd be able to reroute a request meant for org A with the header and the Role header would change their org A role) by "just" setting that org ID. I brought this concern up in https://github.com/grafana/grafana/pull/33162#issuecomment-828433358 where it wasn't applicable, but here it seems to be a concern since the org ID header used here isn't inactive by default.

That PR was ultimately closed by @torkelo in https://github.com/grafana/grafana/pull/34261#pullrequestreview-669073038 because we're trying to keep new multi-org features as part of enterprise. To me, this is on the edge of that comment's reasoning.

Could we change this PR to drop `X-Grafana-Org-Id` support and just set the role for the default org? Something like

```suggestion
					orgID := int64(1)
					if setting.AutoAssignOrg && setting.AutoAssignOrgId > 0 {
						orgID = int64(setting.AutoAssignOrgId)
					}
					extUser.OrgRoles[orgID] = rt
```

It would speed up getting this merged significantly.

Reply:
Thanks @sakjur! Updated to only apply the role to the default org as suggested.